### ISSUE TITLE
samsung-p4wifi kernel 4.18-rc1 is not booting past switch_root 

### ISSUE THREAD

**Issue by Decatf** [postmarketOS/pmbootstrap#1582]
The samsung-p4wifi kernel 4.17 and prior are working. I've rebased onto 4.18-rc1. It isn't booting past `switch_root`. The ssh shell isn't accessible so it hasn't booted that far. 

In attempt to get some more logs. I changed the init script to log to kmsg just before the switch_root:
```
echo "Performing switch_root now." | tee /dev/kmsg

# shellcheck disable=SC2093
exec switch_root /sysroot /sbin/init

echo "ERROR: switch_root failed!" | tee /dev/kmsg
echo "Looping forever. Install and use the debug-shell hook to debug this." | tee /dev/kmsg
echo "For more information, see <https://postmarketos.org/debug-shell>" | tee /dev/kmsg
loop_forever
```

There's the log right before switch_root. There isn't any `ERROR: switch_root failed` message. So I either it doesn't get beyond that command or maybe it stuck there?
```
[   22.791384] Performing switch_root now.
```

I don't know out how to do further debugging of the boot process between the `switch_root` and ssh shell becoming available.

**Comment by ollieparanoid** [comment]
Thanks for the detailed report.

Some ideas:
* Would it be feasible for you to do a regression test of the kernel?
* Maybe `switch_root` prints out anything helpful, so adding `| tee /dev/kmsg` after it might help?
* For the lack of a better idea, consider adding debug code to `switch_root.c` and executing that instead?

See also:
* [switch_root.c](https://github.com/mirror/busybox/blob/master/util-linux/switch_root.c) (there's a long comment at the end of the file describing how it works)

**Comment by Decatf** [comment]
I've got it booting. There was no apparent errors from `switch_root` so I looked into getting the init.d scripts to log to kmsg:

1. Add the following to `/etc/init.d/functions.sh` so that the services will log to `kmsg`.
```shell
[ -w /dev/kmsg ] && exec 30>/dev/kmsg && BASH_XTRACEFD=30
set -x
```
[1]https://bugzilla.redhat.com/show_bug.cgi?id=1168664#c5
[2]https://github.com/fedora-sysv/initscripts/blob/218e4d58ed0a1a8a5e6d75b178940a097444801d/rc.d/init.d/functions#L819-L820

2. Disable kernel printk ratelimit using this cmdline option `printk.devkmsg=on`.

3. The last kmsg becomes extremely verbose so I looked for the service names:
```shell
$ cat /sys/fs/pstore/console-ramoops-0 | grep SVCNAME=
[    8.733902] SVCNAME=sysfs
[    8.914076] SVCNAME=udev
[    9.138345] SVCNAME=udev-trigger
[    9.954524] SVCNAME=modules
[   10.521438] SVCNAME=hwclock
```

4. The last service to run was `hwclock`. I enabled the `tegra20-rtc` device and the `hwclock` service does not hang anymore.
The p4wifi has two clock sources. One provided by CPU and one by PMIC. The CPU RTC suspends which caused some trouble on Android with clock going out of sync. That is why I had disabled that driver.

That's how I traced this part of the boot process. There's a potential cause of the problem with [these](https://lwn.net/Articles/757403/) clock source patches in 4.18. Haven't had time to read into that yet.
